yt/data_objects/data_containers.py

```diff
@@ -2179,3 +2182,93 @@ def _reconstruct_object(*args, **kwargs):
     obj = cls(*new_args)
     obj.field_parameters.update(field_parameters)
     return ReconstructedObject((ds, obj))
+
+def monte_carlo_sample(ds, n_samples=100000, fields=None, n_neighbors=1):
+    """
+    Uses Monte Carlo sampling to create a particle-based dataset out of a
+    grid-based dataset.  For each sample, the code probabilistically selects
+    a cell weighted according to its mass, and a particle is created possessing
+    the gas fields of the cell at some random location within the cell's volume.
+
+    By default the fields that are created are mass, density, smoothing_length,
+    and the position and velocity fields.
+
+    Parameters
+    ----------
+    ds : Dataset
+        Grid-based dataset used as source
+
+    n_samples : int
+        Number of times to sample the dataset
+
+    fields : list
+        List of additional fields you want to include in returned dataset
+
+    n_neighbors : int
+        Number of neighbors for each node in kdtree to determine
+        smoothing_length
+
+    Returns
+    -------
+    StreamParticlesDataset
+
+    Examples
+    --------
+    >>> ds_grid = yt.load('IsolatedGalaxy/galaxy0030/galaxy0030')
+    >>> ds_sph = monte_carlo_sample(ds_grid)
+    """
+    from yt.frontends.stream.api import \
+        load_particles
+    l_unit = 'code_length'
+    m_unit = 'code_mass'
+    d_unit = 'code_mass/code_length**3'
+    v_unit = 'code_length/code_time'
+    rands = np.random.rand(n_samples)
+    m_tot = ds.r[('gas', 'cell_mass')].sum()
```

Comment:
If this gets turned into something that samples a data object, this won't be a relevant comment, but I would suggest instead of using `ds.r[tuple]`, you make a new data object by doing `ds.all_data()` -- when queried in that way, `ds.r[]` will re-use the same data object, so the fields will stick around.

Reply:
That's fair.  I wasn't sure if that would unnecessarily allocate all of `ad`s fields and use up memory.  But I think you're right that this is the way to go.

Reply:
If it does end up as a standalone function it'll need to be added to the API docs manually.

Reply:
I think you're safer with `ad = ds.all_data()` since the fields will not be persistent outside the function.

It does make me wonder if it would be worthwhile to add a derived quantity that does a random sampling.

Reply:
Hmm.  So your suggestion is to make it attached to the derived quantity functionality?  But this isn't really a "quantity" as opposed to an additional data_object, I think.  But I do like the idea of making it work on individual data objects as opposed to datasets as a whole.